### DISCUSSION TITLE
[1.4->1.5] mapping collections now uses `has*` method to determine to set collection value

### DISCUSSION THREAD

**Discussion by marceloverdijk** [mapstruct/mapstruct#3191] [Q&A]
I'm migrating a project using Mapstruct 1.4.3 to 1.5.

Some mappings fail now regarding collections unfortunately and I could not find anything in the breaking changes or the [Collection mapping strategies](https://mapstruct.org/documentation/stable/reference/html/#collection-mapping-strategies) section of the docs.

I have an entity class with:

```java
@Getter
@Setter
private List<CounterEntity> children = new ArrayList<>();

public boolean hasChildren() {
    return children != null && !children.isEmpty();
}
```

and a DTO containing:

```java
private List<CounterDTO> children; // note not initialized
```

with 1.4.3 this generates something like:

```java
@Override
public CounterDTO mapFromDomain(CounterEntity entity) {
    ...
    counterDTO.setChildren( mapFromDomain( entity.getChildren() ) );
    ...
}

@Override
public List<CounterDTO> mapFromDomain(List<CounterEntity> entities) {
    List<CounterDTO> target = hibernateAwareMapper.fixLazyList( entities, List.class );
    if ( target != null ) {
        return target;
    }

    if ( entities == null ) {
        return null;
    }

    List<CounterDTO> list = new ArrayList<CounterDTO>( entities.size() );
    for ( CounterEntity counterEntity : entities ) {
        list.add( mapFromDomain( counterEntity ) );
    }

    return list;
}
```

in this case the `children` property in the DTO will be an empty list if the entity's property was an empty list.

But from 1.5 it now generates:

```java
if ( entity.hasChildren() ) {
    counterDTO.setChildren( mapFromDomain( entity.getChildren() ) );
}
```

so it now uses the `hasChildren` method available in the source entity class to determine whether setting the children or not.

So in case the entity's children is an empty list, the mapper will not set the DTO children, which will stay `null` in that case.

I this *new* behaviour configurable or what is my best option to solve this?

**Comment by filiphr** [comment]
The reason for this change is because what was happening in 1.4 was a bug. See #2423.

If you change your example a bit and add `CounterEntity child` with a `public boolean hasChild()` method you will see that MapStruct is going to call that presence check method (`hasChild`) before doing the mapping. The fact that it was. not being called for collections was a bug.
